### PR TITLE
[SPARK-24772][SQL] Avro: support logical date type

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Support Avro logical date type:
https://avro.apache.org/docs/1.8.2/spec.html#Date

## How was this patch tested?

Unit test

### CODE REVIEW

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala [hunk 1]
```diff
@@ -92,7 +92,7 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
       case BinaryType =>
         (getter, ordinal) => ByteBuffer.wrap(getter.getBinary(ordinal))
       case DateType =>
-        (getter, ordinal) => getter.getInt(ordinal) * DateTimeUtils.MILLIS_PER_DAY
+        (getter, ordinal) => getter.getInt(ordinal)
```
Comment (gengliangwang) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: For the write path, let's drop the previous conversion to `Long`

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: Does this cause a behaviour change?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala [hunk 2]
```diff
@@ -100,6 +103,8 @@ class AvroDeserializer(rootAvroType: Schema, rootCatalystType: DataType) {
           s"Cannot convert Avro logical type ${other} to Catalyst Timestamp type.")
       }
 
+      // Before we upgrade Avro to 1.8 for logical type support, spark-avo converts Long to Date.
```
Comment (viirya) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala]: typo: spark-avo -> spark-avro.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala (hunk 1 above)
Comment (cloud-fan) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: There are 2 kinds of compatibilities:
1. the file written by old avro data source can be read by the new avro data source
2. the file written by new avro data source can be read by the old avro data source

I think we should focus on 1) and ignore 2)

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Comment by HyukjinKwon** [comment]
LGTM too

**Pull request closed without merging**